core/src/main/scala/kafka/log/Log.scala

```diff
@@ -345,13 +346,8 @@ class Log(@volatile var dir: File,
   }
 
   private def recoverLog() {
-    // if we have the clean shutdown marker, skip recovery
-    if(hasCleanShutdownFile) {
```

Comment:
Hmm, this can be a bit tricky. During a clean shutdown, we will flush all segments in a log and write the nextOffset as the flush point in the checkpoint file. The issue with the patch is that logSegments(this.recoveryPoint, Long.MaxValue) will return the last segment during. This forces us to read the last segment of every log even in a clean shutdown, which will be expensive. 

We could potentially improve this by passing in recoveryPoint to recoverSegment() and only read the log from the recovery point. Finding the position in the log segment requires accessing the offset index. However, if the shutdown is unclean, the offset index may not be in a usable state until after the recovery is done.